### ISSUE TITLE
Presigned URL resolves in a SignatureDoesNotMatch

### ISSUE THREAD

**Issue by TVDSC** [aws/aws-sdk-java-v2#2154]
## Describe the issue

Hi,
I have functionalities in our web application to upload and download a certain configuration file (xml).
The upload works.
But we want to enforce the download by using a presigned URL that is valid for 5 minutes.

When using that url we receive a SignatureDoesNotMatch.
<img width="1632" alt="Screenshot 2020-11-20 at 11 00 09" src="https://user-images.githubusercontent.com/26546590/99787278-d5730e80-2b1f-11eb-995c-113a1a3f5b68.png">

## Steps to Reproduce
> Code to upload a configuration file:
```
public Boolean uploadConfiguration(Part file) throws IOException {
        String bucketName = System.getenv("AWS_BUCKET");
        PutObjectRequest request = PutObjectRequest.builder()
                .bucket(bucketName)
                .key("BeFirstConfiguration.xml")
                .build();
        PutObjectResponse response = s3Client.putObject(request, RequestBody.fromBytes(file.getInputStream().readAllBytes()));
        return StringUtils.isNotEmpty(response.eTag());
    }
```

> Code to receive a presigned URL:
```
String bucketName = System.getenv("AWS_BUCKET");

        GetObjectRequest getUrlRequest = GetObjectRequest.builder()
                .bucket(bucketName)
                .key("BeFirstConfiguration.xml")
                .build();
        GetObjectPresignRequest getObjectPresignRequest = GetObjectPresignRequest.builder()
                .getObjectRequest(getUrlRequest)
                .signatureDuration(Duration.ofMinutes(10))
                .build();

        PresignedGetObjectRequest request = s3Presigner.presignGetObject(getObjectPresignRequest);

        log.info("Generating pre-signed URL.");
        return request.url().toString();
```

> JavaConfig configuration
```
public S3Configuration s3Configuration() {
        return S3Configuration.builder().build();
    }

    @Bean
    public S3Presigner s3Presigner() {
        return S3Presigner.builder()
                .region(Region.EU_WEST_1)
                .serviceConfiguration(s3Configuration())
                .build();
    }

    @Bean
    public S3Client s3Client() {
        return S3Client.create();
    }
```

## Current Behavior

After using that presigned url we get an SignatureDoesNotMatch.

## Your Environment

* AWS Java SDK version used:  2.15.31
* JDK version used: openjdk:11 (docker image based on openjdk:11-jre-slim)
* Operating System and version: Amazon Linux

Any ideas or help pls? Been stuck here for a while now.

kind regards

**Comment by duboisph** [comment]
Hmm, not sure whether we can help much. However there's one thing I can think of:

Internally we use private VPC Endpoints for S3 (and DynamoDB). So maybe that's conflicting. Can you specify which endpoint to use in your signing code?

For example with the AWS CLI that's possible: `aws s3 presign --endpoint-url https://s3.eu-west-1.amazonaws.com s3://MY_BUCKET_NAME/MY_FILE.pdf --region eu-west-1`


![image](https://user-images.githubusercontent.com/33081/99789345-b4f88380-2b22-11eb-82be-9993a09bc755.png)

**Comment by TVDSC** [comment]
I can add an endpoint override to the `S3Presigner`.
Which i tried with the value `https://s3.eu-west-1.amazonaws.com`.
```
@Bean
    public S3Presigner s3Presigner() {
        return S3Presigner.builder()
                .region(Region.EU_WEST_1)
                .serviceConfiguration(s3Configuration())
                .endpointOverride(URI.create("https://s3.eu-west-1.amazonaws.com"))
                .build();
    }
```

But then i get the same result.

**Comment by TVDSC** [comment]
Hmm do we perhaps need to execute `putObject` with an `PresignedPutObjectRequest` instead of `PutObjectRequest` i'm using now ?

**Comment by duboisph** [comment]
Then I'm at a loss, don't know how to use the java (or other) SDKs :sweat: . I tried the awscli on the graphql container and that works:

```
> kubectl exec -it -n development graphql-service-dev-6ccb5fb69b-c9gxk -- /bin/bash

root@graphql-service-dev-6ccb5fb69b-c9gxk:/# apt update && apt install -y awscli
<snip>

root@graphql-service-dev-6ccb5fb69b-c9gxk:/# aws s3 presign --endpoint-url https://s3.eu-west-1.amazonaws.com s3://neanex-uploads-staging-development/BeFirstConfiguration.xml --region eu-west-1
https://s3.eu-west-1.amazonaws.com/neanex-uploads-staging-development/BeFirstConfiguration.xml?AWSAccessKeyId=REDACTED&Signature=REDACTED&Expires=1605872947
```

Downloading the file with that URL in my browser works.

**Comment by TVDSC** [comment]
Yes for me as well.
Anyone else that has any idea why it doesn't work with the SDK?

**Comment by TVDSC** [comment]
<img width="1619" alt="Screenshot 2020-11-20 at 12 50 39" src="https://user-images.githubusercontent.com/36960050/99797325-55ed3b80-2b2f-11eb-82f7-9fed014f19d8.png">

The urls are different as well.
Left is using the code example from the first post.
Right is using @duboisph aws cli presign comman.

**Comment by duboisph** [comment]
Could it be you're generating a presigned request to upload a file with the code, instead of downloading?

The example in https://docs.aws.amazon.com/AmazonS3/latest/dev/ShareObjectPreSignedURLJavaSDK.html looks different than what you're trying.

```
            GeneratePresignedUrlRequest generatePresignedUrlRequest =
                    new GeneratePresignedUrlRequest(bucketName, objectKey)
                            .withMethod(HttpMethod.GET)
                            .withExpiration(expiration);
            URL url = s3Client.generatePresignedUrl(generatePresignedUrlRequest);
```

**Comment by TVDSC** [comment]
Nobody that has experience or has worked with these Presigned requests?

**Comment by debora-ito** [comment]
Hi @TVDSC sorry for the delay in response. The object does not need to be uploaded with a presigned url to be downloaded with a presigned url. 

In the url on the left ^^^ `x-amz-te` is listed as a signed header but no value is sent for this header in the query params, I think this is the root cause. 

I see the object you've uploaded is a Part, is it one single S3 [Part](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/model/Part.html)? I'm trying to pinpoint why the presigner would add `x-amz-te` to the list of signerd headers.

**Comment by TVDSC** [comment]
Hi @debora-ito Thanks for your reply.

This is an `javax.servlet.http.Part` object, not from the AWS SDK.
`import javax.servlet.http.Part;` 

Could that be the issue? Although I'm just reading the bytes.

kind regards

**Comment by TVDSC** [comment]
@debora-ito Any more input you can provide?

**Comment by debora-ito** [comment]
@TVDSC Are you enabling checksum validation anywhere in the code? It's one way I could think of that would add the `x-amz-te` header to the signed headers list.

Are you using a proxy?

**Comment by TVDSC** [comment]
This issue is not resolved.

**Comment by debora-ito** [comment]
@TVDSC Okay, I can reproduce the issue now. I was testing my theory to see if the error only happens when checksum validation is enabled in the S3Presigner serviceConfiguration, but found out that setting an empty S3Configuration also adds the `x-amz-te` header.

Can you test this code based in your sample, and see if you url succeeds?
```
public S3Configuration s3Configuration() {
        return S3Configuration.builder().build();
    }

    @Bean
    public S3Presigner s3Presigner() {
        return S3Presigner.builder()
                .region(Region.EU_WEST_1)
//                .serviceConfiguration(s3Configuration())  ---> comment this line
                .build();
    }
```

**Comment by TVDSC** [comment]
@debora-ito Thx! I will and let you know asap.

**Comment by debora-ito** [comment]
@TVDSC I thought this was a bug but actually it is not. Some presigned urls cannot be executed by a web browser, and the way to determine this is by running the `PresignedRequest.isBrowserExecutable()` flag.

The [S3Presigner](http://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/presigner/S3Presigner.html) javadoc has a couple of examples showing how to execute the presigned urls programmatically, I ran for the url generated by your sample code and it works fine. 

Running your sample code without setting `.serviceConfiguration()` will generate a url that is browser compatible, so this is another option.

**Comment by neanex** [comment]
hi @debora-ito I try to have a look this week and let you know.
Thanks for this info!

kind regards

**Comment by TVDSC** [comment]
hi @debora-ito I try to have a look this week and let you know.
Thanks for this info!

kind regards

**Comment by tuckerbuchy** [comment]
@debora-ito  curious why this is not a bug? S3Configuration seems to have a lot more uses, such as using a transfer accelerated host name. I'm having a lot of trouble getting this to work because I have to use 

```
S3Configuration s3Configuration = S3Configuration.builder().accelerateModeEnabled(true).build();
S3Presigner customS3Presigner = S3Presigner.build().serviceConfiguration(s3Configuration).build();
```

My READ requests all now contain the x-amz-te header, causing signature failures. Is there something I am missing?

**Comment by impalashag** [comment]
Hi @TVDSC,

I had the similar issue. I was able to generate the presigned url for an object in s3 however the url was giving the same error while accessing it on browser. In my case, it had nothing to do with checksum validations because when I checked the PresignedRequest.isBrowserExecutable() flag it was always returning true.

After digging a lot I found out I was not providing the correct bucket name and file path. My file was inside a directory/folder in a bucket and I was appending this directory with the bucket name itself instead of file name and hence the error. If you see the error message, it clearly states "Check your key and signing method". Hope that helps.

cc @debora-ito
